src/aws_encryption_sdk/keyring/raw_keyring.py

```diff
@@ -0,0 +1,390 @@
+        _password = attr.ib(validator=attr.validators.optional(attr.validators.instance_of(six.binary_type)))
+        _wrapping_algorithm = attr.ib(repr=False, validator=attr.validators.instance_of(WrappingAlgorithm))
+        if _password:
+            loaded_wrapping_key = serialization.load_der_private_key(
+                data=_encoded_key, password=_password, backend=default_backend()
+            )
+        else:
+            loaded_wrapping_key = serialization.load_der_public_key(data=_encoded_key, backend=default_backend())
+        return RawRSAKeyring(
+            key_namespace=key_namespace,
+            key_name=key_name,
+            wrapping_key=loaded_wrapping_key,
+            wrapping_algorithm=_wrapping_algorithm,
+        )
+
+    def __attrs_post_init__(self):
+        # type: () -> None
+        """Prepares initial values not handled by attrs."""
+        self._key_provider = MasterKeyInfo(provider_id=self.key_namespace, key_info=self.key_name)
+
+    def on_encrypt(self, encryption_materials):
+        # type: (EncryptionMaterials) -> EncryptionMaterials
+        """Generate a data key if not present and encrypt it using any available wrapping key.
+
+        :param encryption_materials: Encryption materials for the keyring to modify.
+        :type encryption_materials: aws_encryption_sdk.materials_managers.EncryptionMaterials
+        :returns: Optionally modified encryption materials.
+        :rtype: aws_encryption_sdk.materials_managers.EncryptionMaterials
+        """
+        if not encryption_materials.data_encryption_key:
+            plaintext_data_key = generate_data_key(
+                encryption_materials=encryption_materials, key_provider=self._key_provider
+            )
+        else:
+            plaintext_data_key = encryption_materials.data_encryption_key.data_key
+        if isinstance(self._wrapping_key, rsa.RSAPublicKey):
+
+            # Encrypt data key
+            encrypted_wrapped_key = EncryptedData(
+                iv=None,
+                ciphertext=self._wrapping_key.encrypt(
+                    plaintext=plaintext_data_key, padding=self._wrapping_algorithm.padding
+                ),
+                tag=None,
+            )
+        else:
+            # Encrypt data key
+            encrypted_wrapped_key = EncryptedData(
+                iv=None,
+                ciphertext=self._wrapping_key.public_key().encrypt(
+                    plaintext=plaintext_data_key, padding=self._wrapping_algorithm.padding
+                ),
+                tag=None,
+            )
+
+        # EncryptedData to EncryptedDataKey
+        encrypted_data_key = serialize_wrapped_key(
+            key_provider=self._key_provider,
+            wrapping_algorithm=self._wrapping_algorithm,
+            wrapping_key_id=self.key_name,
+            encrypted_wrapped_key=encrypted_wrapped_key,
+        )
+
+        # Update Keyring Trace
+        if encrypted_data_key:
```

Comment:
What is the logic flow that would result in this being false but also not have thrown an error?